src/deluge/gui/views/timeline_view.cpp

```diff
@@ -146,24 +148,36 @@ ActionResult TimelineView::horizontalEncoderAction(int32_t offset) {
 
 		if (isUIModeActiveExclusively(UI_MODE_HOLDING_HORIZONTAL_ENCODER_BUTTON)) {
 			int32_t oldXZoom = currentSong->xZoom[navSysId];
-
 			int32_t zoomMagnitude = -offset;
+			uint32_t newZoom = zoomMagnitude == -1 ? oldXZoom >> 1 : oldXZoom << 1;
+			uint32_t clipLengthMaxZoom = getMaxZoom();
 
-			// Constrain to zoom limits
-			if (zoomMagnitude == -1) {
-				if (currentSong->xZoom[navSysId] <= 3) {
+			// When this timer is set, we may need to wait a bit before allowing zooming
+			if (timeHorizontalZoomLastHitClipMax != 0) {
+				bool isWithinZoomDelay = util::infinite_a_lt_b(
+				    AudioEngine::audioSampleTimer - timeHorizontalZoomLastHitClipMax, kShortPressTime);
+
+				// Prevent further zooming in the delayed direction (see below)
+				if (isWithinZoomDelay && (delayedHorizontalScrollDirection == zoomMagnitude)) {
 					goto getOut;
 				}
-				currentSong->xZoom[navSysId] >>= 1;
+
+				timeHorizontalZoomLastHitClipMax = 0;
 			}
-			else {
-				if (currentSong->xZoom[navSysId] >= getMaxZoom()) {
-					goto getOut;
-				}
-				currentSong->xZoom[navSysId] <<= 1;
+
+			// Prevent quickly scrolling past the clip's max zoom level
+			if (newZoom == clipLengthMaxZoom && timeHorizontalZoomLastHitClipMax == 0) {
+				timeHorizontalZoomLastHitClipMax = AudioEngine::audioSampleTimer;
```

Comment:
I think the `kShortPressTime` should be added here and then ClipMax used as a deadline, so we can compare it directly to the audio sample timer above rather than redoing math.

Reply:
Updated this along with the variable naming to be more clear (I think). Definitely open to further refinements though.

Reply:
looks good, thanks!